src/dbnode/environment/config_test.go

```diff
@@ -30,50 +32,370 @@ import (
+		},
+		{
+			numHosts:  5,
+			numShards: 16,
+			rf:        3,
+			expectErr: false,
+		},
+		{
+			numHosts:  100,
+			numShards: 4096,
+			rf:        3,
+			expectErr: false,
+		},
+		{
+			numHosts:  2,
+			numShards: 16,
+			rf:        3,
+			expectErr: true,
+		},
+		{
+			numHosts:  0,
+			numShards: 16,
+			rf:        3,
+			expectErr: true,
+		},
+		{
+			numHosts:  10,
+			numShards: 16,
+			rf:        0,
+			expectErr: true,
+		},
+		{
+			numHosts:  10,
+			numShards: 0,
+			rf:        3,
+			expectErr: true,
+		},
+		{
+			numHosts:  10,
+			numShards: 16,
+			rf:        11,
+			expectErr: true,
+		},
+	}
+
+	for _, test := range tests {
+		validity := "valid"
+		if test.expectErr {
+			validity = "invalid"
+		}
+		testName := fmt.Sprintf("%s:%dhosts;%drf;%dshards", validity, test.numHosts, test.rf, test.numShards)
+		t.Run(testName, func(t *testing.T) {
+			var hosts []topology.HostShardConfig
+			for i := 0; i < test.numHosts; i++ {
+				hosts = append(hosts, topology.HostShardConfig{
+					HostID:        fmt.Sprintf("id%d", i),
+					ListenAddress: fmt.Sprintf("id%d", i),
+				})
+			}
+
+			hostShardSets, err := generatePlacement(hosts, test.numShards, test.rf)
+			if test.expectErr {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+
+			var (
+				minShardCount = test.numShards + 1
+				maxShardCount = 0
+				shardCounts   = make([]int, test.numShards)
+			)
+			for _, hostShardSet := range hostShardSets {
+				ids := hostShardSet.ShardSet().AllIDs()
+				if len(ids) < minShardCount {
+					minShardCount = len(ids)
+				}
+				if len(ids) > maxShardCount {
+					maxShardCount = len(ids)
+				}
+
+				for _, id := range ids {
+					shardCounts[id]++
+				}
+			}
+
+			// Assert balanced shard distribution
+			assert.True(t, maxShardCount-minShardCount < 2)
+			// Assert each shard has `rf` replicas
+			for _, shardCount := range shardCounts {
+				assert.Equal(t, test.rf, shardCount)
+			}
+		})
+	}
+}
+
+func TestGeneratePlacementConsistency(t *testing.T) {
+	// Asserts that the placement generated by `generatePlacement` is
+	// deterministic even when the ordering of hosts passed in is in a
+	// different order.
+	var (
+		numHosts  = 123
+		numShards = 4096
+		rf        = 3
+		iters     = 10
+		hosts     = make([]topology.HostShardConfig, 0, numHosts)
+	)
+
+	for i := 0; i < numHosts; i++ {
+		hosts = append(hosts, topology.HostShardConfig{
+			HostID:        fmt.Sprintf("id%d", i),
+			ListenAddress: fmt.Sprintf("id%d", i),
+		})
+	}
+
+	var pl []topology.HostShardSet
+	for i := 0; i < iters; i++ {
+		rand.Shuffle(len(hosts), func(i, j int) {
+			hosts[i], hosts[j] = hosts[j], hosts[i]
+		})
+
+		hostShardSets, err := generatePlacement(hosts, numShards, rf)
+		require.NoError(t, err)
+
+		if i == 0 {
+			pl = hostShardSets
+		} else {
+			assertHostShardSetsEqualish(t, pl, hostShardSets)
+		}
+	}
+}
+
+func assertHostShardSetsEqualish(t *testing.T, one, two []topology.HostShardSet) {
```

Comment:
Lol "equalish" :D . Add a comment describing the "ish" part? i.e., which parts are exactly equal, which parts are fuzzy.